File: frontend/src/services/auth.service.js
import axios from 'axios';

const API_URL = process.env.REACT_APP_API_URL || 'https://game-ujiz.onrender.com/api';

/**
 * Auth Service
 * Handles API calls related to authentication
 */
class AuthService {
  /**
   * Register a new user
   * @param {string} username - Username
   * @param {string} email - Email
   * @param {string} password - Password
   * @param {Object} profile - Additional profile data
   * @returns {Promise} - Promise with registration result
   */
  register(username, email, password, profile = {}) {
    return axios.post(`${API_URL}/auth/register`, {
      username,
      email,
      password,
      ...profile
    });
  }

  /**
   * Login user
   * @param {string} username - Username or email
   * @param {string} password - Password
   * @returns {Promise} - Promise with login result
   */
  login(username, password) {
    return axios
      .post(`${API_URL}/auth/login`, {
        username,
        password
      })
      .then(response => {
        if (response.data.data.token) {
          localStorage.setItem('user', JSON.stringify(response.data.data));
        }

        return response.data;
      });
  }

  /**
   * Logout user
   */
  logout() {
    localStorage.removeItem('user');
  }

  /**
   * Get current user
   * @returns {Object} - Current user data
   */
  getCurrentUser() {
    return JSON.parse(localStorage.getItem('user'));
  }

  /**
   * Check if user is logged in
   * @returns {boolean} - True if user is logged in
   */
  isLoggedIn() {
    const user = this.getCurrentUser();
    return !!user && !!user.token;
  }

  /**
   * Refresh token
   * @returns {Promise} - Promise with refresh result
   */
  refreshToken() {
    const user = this.getCurrentUser();
    
    if (!user || !user.refreshToken) {
      return Promise.reject('No refresh token available');
    }
    
    return axios
      .post(`${API_URL}/auth/refresh-token`, {
        refreshToken: user.refreshToken
      })
      .then(response => {
        if (response.data.data.token) {
          // Update stored user with new tokens
          localStorage.setItem('user', JSON.stringify({
            ...user,
            token: response.data.data.token,
            refreshToken: response.data.data.refreshToken
          }));
        }
        
        return response.data;
      });
  }

  /**
   * Update user profile
   * @param {Object} profileData - Profile data to update
   * @returns {Promise} - Promise with update result
   */
  updateProfile(profileData) {
    const user = this.getCurrentUser();
    
    if (!user || !user.token) {
      return Promise.reject('User not authenticated');
    }
    
    return axios
      .put(
        `${API_URL}/users/profile`, 
        profileData,
        { headers: { Authorization: `Bearer ${user.token}` } }
      )
      .then(response => {
        // Update stored user with new profile data
        localStorage.setItem('user', JSON.stringify({
          ...user,
          ...response.data.data.user
        }));
        
        return response.data;
      });
  }

  /**
   * Request password reset
   * @param {string} email - User email
   * @returns {Promise} - Promise with request result
   */
  requestPasswordReset(email) {
    return axios.post(`${API_URL}/auth/forgot-password`, { email });
  }

  /**
   * Reset password
   * @param {string} token - Reset token
   * @param {string} password - New password
   * @returns {Promise} - Promise with reset result
   */
  resetPassword(token, password) {
    return axios.post(`${API_URL}/auth/reset-password`, {
      token,
      password
    });
  }
}

export default new AuthService();

File: mobile/lib/src/services/auth_service.dart
import 'dart:convert';
import 'package:flutter_dotenv/flutter_dotenv.dart';
import 'package:http/http.dart' as http;
import 'package:flutter_secure_storage/flutter_secure_storage.dart';

import '../models/user.dart';
import '../utils/api_exception.dart';

class AuthService {
  final String baseUrl;
  final FlutterSecureStorage _storage = const FlutterSecureStorage();
  
  AuthService({String? baseUrl}) : this.baseUrl = baseUrl ?? dotenv.env['API_URL'] ?? 'http://localhost:5000/api';

  // Register a new user
  Future<User> register(String username, String email, String password, String preferredRole) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/auth/register'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({
          'username': username,
          'email': email,
          'password': password,
          'preferredRole': preferredRole,
        }),
      );

      if (response.statusCode == 201) {
        final data = jsonDecode(response.body);
        await _storage.write(key: 'token', value: data['token']);
        return User.fromJson(data['user']);
      } else {
        final error = jsonDecode(response.body);
        throw ApiException(error['message'] ?? 'Registration failed');
      }
    } catch (e) {
      if (e is ApiException) rethrow;
      throw ApiException('Network error: ${e.toString()}');
    }
  }

  // Login user
  Future<User> login(String email, String password) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/auth/login'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({
          'email': email,
          'password': password,
        }),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        await _storage.write(key: 'token', value: data['token']);
        return User.fromJson(data['user']);
      } else {
        final error = jsonDecode(response.body);
        throw ApiException(error['message'] ?? 'Login failed');
      }
    } catch (e) {
      if (e is ApiException) rethrow;
      throw ApiException('Network error: ${e.toString()}');
    }
  }

  // Get current user
  Future<User?> getCurrentUser() async {
    try {
      final token = await _storage.read(key: 'token');
      if (token == null) return null;

      final response = await http.get(
        Uri.parse('$baseUrl/auth/me'),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token',
        },
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return User.fromJson(data['user']);
      } else if (response.statusCode == 401) {
        await _storage.delete(key: 'token');
        return null;
      } else {
        final error = jsonDecode(response.body);
        throw ApiException(error['message'] ?? 'Failed to get user');
      }
    } catch (e) {
      if (e is ApiException) rethrow;
      throw ApiException('Network error: ${e.toString()}');
    }
  }

  // Update user profile
  Future<User> updateProfile(String userId, Map<String, dynamic> updates) async {
    try {
      final token = await _storage.read(key: 'token');
      if (token == null) throw ApiException('Not authenticated');

      final response = await http.put(
        Uri.parse('$baseUrl/users/$userId'),
        headers: {
          'Content-Type': 'application/json',
          'Authorization': 'Bearer $token',
        },
        body: jsonEncode(updates),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return User.fromJson(data['user']);
      } else {
        final error = jsonDecode(response.body);
        throw ApiException(error['message'] ?? 'Failed to update profile');
      }
    } catch (e) {
      if (e is ApiException) rethrow;
      throw ApiException('Network error: ${e.toString()}');
    }
  }

  // Logout user
  Future<void> logout() async {
    await _storage.delete(key: 'token');
  }

  // Check if user is authenticated
  Future<bool> isAuthenticated() async {
    final token = await _storage.read(key: 'token');
    return token != null;
  }

  // Get auth token
  Future<String?> getToken() async {
    return await _storage.read(key: 'token');
  }
}
